cpp/src/io/parquet/page_string_decode.cu

```diff
@@ -0,0 +1,1125 @@
+
+  // need to save num_nulls and num_valids calculated in page_bounds in this page
+  if (t == 0) {
+    pp->num_nulls  = s->page.num_nulls;
+    pp->num_valids = s->page.num_valids;
+  }
+
+  // now process string info in the range [start_value, end_value)
+  // set up for decoding strings...can be either plain or dictionary
+  auto const& col          = s->col;
+  uint8_t const* data      = s->data_start;
+  uint8_t const* const end = s->data_end;
+  uint8_t const* dict_base = nullptr;
+  int dict_size            = 0;
+  size_t str_bytes         = 0;
+
+  switch (pp->encoding) {
+    case Encoding::PLAIN_DICTIONARY:
+    case Encoding::RLE_DICTIONARY:
+      // RLE-packed dictionary indices, first byte indicates index length in bits
+      if (col.str_dict_index) {
+        // String dictionary: use index
+        dict_base = reinterpret_cast<const uint8_t*>(col.str_dict_index);
+        dict_size = col.page_info[0].num_input_values * sizeof(string_index_pair);
+      } else {
+        dict_base = col.page_info[0].page_data;  // dictionary is always stored in the first page
+        dict_size = col.page_info[0].uncompressed_page_size;
+      }
+
+      // FIXME: need to return an error condition...this won't actually do anything
+      if (s->dict_bits > 32 || !dict_base) { CUDF_UNREACHABLE("invalid dictionary bit size"); }
+
+      str_bytes = totalDictEntriesSize(
+        data, dict_base, s->dict_bits, dict_size, (end - data), start_value, end_value);
+      break;
+    case Encoding::PLAIN:
+      dict_size = static_cast<int32_t>(end - data);
+      str_bytes = is_bounds_pg ? totalPlainEntriesSize(data, dict_size, start_value, end_value)
+                               : dict_size - sizeof(int) * (pp->num_input_values - pp->num_nulls);
+      break;
+  }
+
+  if (t == 0) {
+    // TODO check for overflow
+    pp->str_bytes = str_bytes;
+  }
+}
+
+/**
+ * @brief Kernel for computing the string column data stored in the pages
+ *
+ * This function will write the page data and the page data's validity to the
+ * output specified in the page's column chunk.
+ *
+ * This version uses a single warp to do the string copies.
+ *
+ * @param pages List of pages
+ * @param chunks List of column chunks
+ * @param min_row Row index to start reading at
+ * @param num_rows Maximum number of rows to read
+ * @tparam lvl_buf_size Size of the buffer used when decoding repetition and definition levels
+ * @tparam level_t Type used to store decoded repetition and definition levels
+ */
+template <int lvl_buf_size, typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodeStringPageData(
+  PageInfo* pages, device_span<ColumnChunkDesc const> chunks, size_t min_row, size_t num_rows)
+{
+  __shared__ __align__(16) page_state_s state_g;
+  __shared__ __align__(16) page_state_buffers_s state_buffers;
+  __shared__ __align__(4) size_type last_offset;
+
```

Comment:
This data and the rep/def buffers below seem identical to the data in the V2 function.   We should wrap these in a struct that gets declared once (as shared) in the root DecodeStringPageData function and then passed as a reference to various subfunctions.

Reply:
I discussed this offline with @vuule, but the V2 kernel is not currently called, but will likely be the only version once `rle_stream` for decoding is online. I've left it in but `if constexpr`'d out to keep the bit rot away. I'm ok with just getting rid of at this point since there shouldn't be much changing at this point.